### ISSUE TITLE
trace_stores() have an impact on pipeline

### ISSUE THREAD

**Issue by Barkar19** [halide/Halide#3812]
Hi everyone,

probably  I have found major bug in Halide library. 

TL;DR: function trace_stores() changes output of the pipeline (with trace_stores() everything is correct, without it pipeline generates bad output).

I found dft implementation in Halide (on MIT servers [here](http://groups.csail.mit.edu/graphics/face/xform/halide/Halide/test/performance/fft.cpp)  ). Than I have tried to run it with current Halide version and it works, but... But Halide's placeholders makes something weird with the code. I have tried to debug and realized that, using trace_stores() have an impact on the pipeline result. With tracing everything is ok, without it fails to produce correct output. 

Please see my github repo ( [github](https://github.com/Barkar19/halide-bug) ) for project that demonstrates the problem. Everything you should do is making symbolic link in "ext" directory to halide distribution and run script  ./run.sh for building and running it.  The difference between two compiled versions is here (line 323) of src/fft.cpp

```
    Func fT("T"+std::to_string(i)+"_"+f.name());
    fT(argsT) = f(f.args());
#ifdef TRACE_STORES
    fT.trace_stores();
#endif
    ++i;
```
Please have a look on this issue. I would be vary grateful if you could tell me what is going on or what I am doing wrong. Thanks!
